ldap/servers/slapd/ldapi.c

```diff
@@ -0,0 +1,408 @@
+                            NULL, entries[0]);
+                    }
+                }
+            }
+
+        entry_map_free:
+            /* auth_dn consumed by bind creds set */
+            slapi_free_search_results_internal(search_pb);
+            slapi_pblock_destroy(search_pb);
+            slapi_ch_free_string(&filter);
+            slapi_ch_free_string(&utype);
+            slapi_ch_free_string(&gtype);
+            slapi_ch_free_string(&base_dn);
+        }
+
+        /*
+         * We map the current process uid also to directory manager.
+         * This is secure as it requires local machine OR same-container volume
+         * access and the correct uid access. If you have access to the uid/gid
+         * and are on the same machine you could always just reset the rootdn hashes
+         * anyway ... so this is no reduction in security.
+         */
+
+        if (ret && (0 == uid || proc_uid == uid || proc_gid == gid)) {
+            /* map unix root (uidNumber:0)? */
+            char *root_dn = config_get_ldapi_root_dn();
+
+            if (root_dn) {
+                Slapi_PBlock *entry_pb = NULL;
+                Slapi_DN *edn = slapi_sdn_new_dn_byref(slapi_dn_normalize(root_dn));
+                Slapi_Entry *e = NULL;
+
+                /* root might be locked too! :) */
+                ret = slapi_search_get_entry(&entry_pb, edn, 0, &e, (void *)plugin_get_default_component_id());
+                if (0 == ret && e) {
+                    ret = slapi_check_account_lock(
+                        0, /* pb not req */
+                        e,
+                        0, /* no response control */
+                        0, /* don't check password policy */
+                        0  /* don't send ldap result */
+                        );
+
+                    if (1 == ret)
+                        /* sorry root,
+                         * just not cool enough
+                         */
+                        goto root_map_free;
+                }
+
+                /* it's ok not to find the entry,
+                 * dn doesn't have to have an entry
+                 * e.g. cn=Directory Manager
+                 */
+                bind_credentials_set_nolock(
+                    conn, SLAPD_AUTH_OS, root_dn,
+                    NULL, NULL, NULL, e);
+
+            root_map_free:
+                /* root_dn consumed by bind creds set */
+                slapi_sdn_free(&edn);
+                slapi_search_get_entry_done(&entry_pb);
+                ret = 0;
+            }
+        }
+
+#if defined(ENABLE_AUTO_DN_SUFFIX)
+        if (ret) {
+            /* create phony auth dn? */
+            char *base = config_get_ldapi_auto_dn_suffix();
+            if (base) {
+                char *tpl = "gidNumber=%u+uidNumber=%u,";
+                int32_t len =
+                    strlen(tpl) +
+                    strlen(base) +
+                    51 /* uid,gid,null,w/padding */
```

Comment:
Does this account for 32bit uid/gid values rather than just 16 bit ones?

Reply:
No idea, this is all copy & paste code.  I will need to look into it...

Reply:
Yeah, I think we should do our due diligence here :)

Reply:
^---